### PR TITLE
Add ram settings validation and flags to disable it

### LINKED ISSUE #29: ram_max_used fix ineffective on updated devices due to persistent storage
As the /usr/app/boinc folder became a volume, the changes fixing #25 do not come to life on existing instances:
Screenshot from 2020-04-24 18-10-34

Only one of the global override files was edited, but this seems to not matter.

### PR DESCRIPTION
**Pull request by ptrm**
Fixes #29

There's also check for `SKIP_BOINC_MEM_SETTINGS_CHECK` and `SKIP_BOINC_CPU_SETTINGS_CHECK` environment variables to disable the auto-correction of config values. 

Adds ca. 5MB to boinc-client image because of adding xml-twig-tools to read config values.

### CODE REVIEW

**Comment by ptrm** [comment]
[This comment from the related issue](https://github.com/balenalabs/rosetta-at-home/issues/29#issuecomment-619569560) is also relevant here

File: boinc-client/start-utils.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  sed -i -e "s|<$1>[0-9a-z.]\{1,\}</$1>|<$1>$2.000000</$1>|g" "$3"
+}
+
+validate_ram_settings() {
+  local cfg_ram_max_busy=$(get_int_xml_val "$cfg_ram_max_busy_xml_key" "$prefs_file_path")
+  local cfg_ram_max_idle=$(get_int_xml_val "$cfg_ram_max_busy_xml_key" "$prefs_file_path")
+
+  echo "Validating boinc RAM settings"
+
+  if [[ ! -z $cfg_ram_max_busy && $cfg_ram_max_busy -gt $treshold_ram_settings_pct ]]; then
```
Comment (chrisys) [inline boinc-client/start-utils.sh]: Being nitpicky, but just wanted to correct spelling here to save any possible mistakes in the future.

```suggestion
  if [[ ! -z $cfg_ram_max_busy && $cfg_ram_max_busy -gt $threshold_ram_settings_pct ]]; then
```

File: boinc-client/start-utils.sh [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    echo "  max RAM when busy (${cfg_ram_max_busy}%) too high - setting to ${treshold_ram_settings_pct}%"
```
Comment (chrisys) [inline boinc-client/start-utils.sh]:
```suggestion
    echo "  max RAM when busy (${cfg_ram_max_busy}%) too high - setting to ${threshold_ram_settings_pct}%"
```

File: boinc-client/start-utils.sh [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    update_float_xml_val_with_int "$cfg_ram_max_busy_xml_key" "$treshold_ram_settings_pct" "$prefs_file_path"
```
Comment (chrisys) [inline boinc-client/start-utils.sh]:
```suggestion
    update_float_xml_val_with_int "$cfg_ram_max_busy_xml_key" "$threshold_ram_settings_pct" "$prefs_file_path"
```

File: boinc-client/start-utils.sh [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  fi
+
+  if [[ ! -z $cfg_ram_max_idle && $cfg_ram_max_idle -gt $treshold_ram_settings_pct ]]; then
```
Comment (chrisys) [inline boinc-client/start-utils.sh]:
```suggestion
  if [[ ! -z $cfg_ram_max_idle && $cfg_ram_max_idle -gt $threshold_ram_settings_pct ]]; then
```

File: boinc-client/start-utils.sh [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    echo "  max RAM when idle (${cfg_ram_max_idle}%) too high - setting to ${treshold_ram_settings_pct}%"
```
Comment (chrisys) [inline boinc-client/start-utils.sh]:
```suggestion
    echo "  max RAM when idle (${cfg_ram_max_idle}%) too high - setting to ${threshold_ram_settings_pct}%"
```

File: boinc-client/start-utils.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    update_float_xml_val_with_int "$cfg_ram_max_idle_xml_key" "$treshold_ram_settings_pct" "$prefs_file_path"
```
Comment (chrisys) [inline boinc-client/start-utils.sh]:
```suggestion
    update_float_xml_val_with_int "$cfg_ram_max_idle_xml_key" "$threshold_ram_settings_pct" "$prefs_file_path"
```

File: boinc-client/start.sh [hunk 7]
```diff
@@ -2,6 +2,15 @@
 mkdir -p /usr/app/boinc/locale
 mkdir -p /usr/app/boinc/slots
 
+prefs_file_path='global_prefs_override.xml'
+cfg_ram_max_busy_xml_key='ram_max_used_busy_pct'
+cfg_ram_max_idle_xml_key='ram_max_used_idle_pct'
+treshold_ram_settings_pct=95
```
Comment (chrisys) [inline boinc-client/start.sh]:
```suggestion
threshold_ram_settings_pct=95
```

File: boinc-client/start.sh [hunk 8]
```diff
@@ -11,14 +20,18 @@ if [[ -z $ACCOUNT_KEY ]]; then
...
+fi
+
+if [[ -z $SKIP_BOINC_MEM_SETTINGS_CHECK ]]; then
+  validate_ram_settings
 fi
 
 totalmem=$(awk '/^MemTotal:/{print $2}' /proc/meminfo)
 
-if [ "$totalmem" -lt "1500000" ]; then
+if [[ -z $SKIP_BOINC_CPU_SETTINGS_CHECK && "$totalmem" -lt "1500000" ]]; then
```
Comment (chrisys) [inline boinc-client/start.sh]: As per your feedback in the issue and my tests on the Pi 4 2GB just now, let's put this back to the 2.5GB value it was at previously.

```suggestion
if [[ -z $SKIP_BOINC_CPU_SETTINGS_CHECK && "$totalmem" -lt "2500000" ]]; then
```

File: boinc-client/start-utils.sh (hunk 1 above)
Comment (ptrm) [reply boinc-client/start-utils.sh]: Absolutely :) Thanks

**Comment by chrisys** [comment]
@balena-ci retest

**Review by chrisys — APPROVED** [review]

File: boinc-client/start.sh [hunk 9]
```diff
@@ -11,14 +20,18 @@ if [[ -z $ACCOUNT_KEY ]]; then
...
+
+if [[ -z $SKIP_BOINC_MEM_SETTINGS_CHECK ]]; then
+  validate_ram_settings
 fi
 
 totalmem=$(awk '/^MemTotal:/{print $2}' /proc/meminfo)
 
-if [ "$totalmem" -lt "1500000" ]; then
+if [[ -z $SKIP_BOINC_CPU_SETTINGS_CHECK && "$totalmem" -lt "2500000" ]]; then
   echo "Less than 1.5GB RAM - running single concurrent task"
```
Comment (rahul-thakoor) [inline boinc-client/start.sh]: I think this should be 2.5GB?

Comment (ptrm) [reply boinc-client/start.sh]: True. fixed

**Comment by rahul-thakoor** [comment]
@balena-ci retest

File: boinc-client/start.sh (hunk 9 above)
Comment (rahul-thakoor) [reply boinc-client/start.sh]: thanks :)

**Review by rahul-thakoor — APPROVED** [review]
Tested working on the pi4 1GB version. Thanks for your contribution :)

**Comment by chrisys** [comment]
+1 thanks for the great contribution @ptrm

**Pull request merged**